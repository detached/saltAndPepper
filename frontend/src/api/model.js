import { ArrayModel, Model } from "objectmodel";

export const FilterKey = {
  AUTHOR: "AUTHOR",
  CATEGORY: "CATEGORY",
  CUISINE: "CUISINE",
};

export const OrderField = {
  CREATED_AT: "CREATED_AT",
  TITLE: "TITLE",
};

export const SortDir = {
  ASC: "ASC",
  DESC: "DESC",
};

export class FilterValue extends Model({
  value: String,
  label: String,
}) {}

export class SearchFilter extends Model({
  AUTHOR: ArrayModel([FilterValue, String]),
  CATEGORY: ArrayModel([FilterValue, String]),
  CUISINE: ArrayModel([FilterValue, String]),
}).defaultTo({
  AUTHOR: [],
  CATEGORY: [],
  CUISINE: [],
}) {}

export class Order extends Model({
  field: [OrderField.CREATED_AT, OrderField.TITLE],
  direction: [SortDir.DESC, SortDir.ASC],
}) {}

export class Page extends Model({
  size: Number,
  number: Number,
  maxNumber: Number,
}).defaultTo({
  maxNumber: 0,
}) {}

export class SearchRequest extends Model({
  searchQuery: String,
  page: Page,
  filter: SearchFilter,
  order: Order,
}) {}

export class SearchResponseData extends Model({
  id: String,
  imageUrl: String,
  title: String,
  category: String,
  cuisine: String,
  author: String,
}) {}

export class SearchResponse extends Model({
  data: ArrayModel(SearchResponseData),
  page: Page,
  possibleFilter: SearchFilter,
}) {}

export class NewRecipeRequest extends Model({
  title: String,
  category: String,
  cuisine: String,
  yields: String,
  ingredients: String,
  instructions: String,
  modifications: String,
}) {}

export class NewRecipeResponse extends Model({
  id: String,
}) {}

export class Profile extends Model({
  id: String,
  name: String,
  role: String,
  isAllowedToInvite: Boolean,
}) {}

export class ChangePasswordRequest extends Model({
  oldPassword: String,
  newPassword: String,
}) {}

export class InvitationCodeResponse extends Model({
  code: String,
}) {}

export class InvitationInfoResponse extends Model({
  invitingUser: String,
}) {}

export class CreateNewCommentRequest extends Model({
  comment: String,
}) {}

export class CommentAuthor extends Model({
  id: String,
  name: String,
}) {}

export class CommentResponse extends Model({
  id: String,
  author: CommentAuthor,
  comment: String,
  createdOn: Date,
  canDelete: Boolean,
}) {}

export class AuthResponse extends Model({
  accessToken: String,
  refreshToken: String,
}) {}
