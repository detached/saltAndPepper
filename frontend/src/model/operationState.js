export const OperationState = {
  NONE: "NONE",
  SUCCESSFUL: "SUCCESSFUL",
  ERROR: "ERROR",
};
