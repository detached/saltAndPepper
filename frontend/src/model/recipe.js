import { ArrayModel, Model } from "objectmodel";

export class RecipeImage extends Model({
  id: String,
  url: String,
  thumbnailUrl: String,
}) {}

export class Author extends Model({
  id: String,
  name: String,
}) {}

export class Recipe extends Model({
  id: [String],
  author: [Author],
  title: String,
  category: String,
  cuisine: String,
  yields: String,
  ingredients: String,
  instructions: String,
  modifications: String,
  images: [ArrayModel(RecipeImage)],
  newImages: [ArrayModel(RecipeImage)],
}) {
  addNewImages(images) {
    if (this.newImages === undefined) {
      this.newImages = [];
    }
    this.newImages = this.newImages.concat(images);
  }

  removeNewImage(imageFile) {
    if (this.newImages === undefined) {
      return;
    }
    this.newImages = this.newImages.filter((image) => image !== imageFile);
  }

  removeExistingImage(imageId) {
    this.images = this.images.filter((image) => image.id !== imageId);
  }
}
