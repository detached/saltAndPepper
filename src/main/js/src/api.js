export function getImagesOfRecipe(recipeId) {
    return fetch('/api/recipe/' + recipeId + '/images');
}

export function addImageToRecipe(recipeId, image) {
    const formData = new FormData();
    formData.append('image', image);
    return fetch('/api/recipe/' + recipeId + '/images', {
        method: 'POST', body: formData
    });
}

export function removeImageFromRecipe(recipeId, imageId) {
    return fetch('/api/recipe/' + recipeId + '/image/' + imageId, {
      method: 'DELETE'
    });
}
