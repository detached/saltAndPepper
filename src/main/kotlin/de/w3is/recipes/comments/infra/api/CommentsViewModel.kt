package de.w3is.recipes.comments.infra.api

import io.micronaut.serde.annotation.Serdeable
import java.time.Instant
import java.util.UUID

@Serdeable
data class CommentsViewModel(
    val recipeId: String,
    val comments: List<CommentViewModel>,
)

@Serdeable
data class CommentViewModel(
    val id: UUID,
    val author: AuthorViewModel,
    val comment: String,
    val createdAt: Instant,
    val canDelete: Boolean,
)

@Serdeable
data class AuthorViewModel(
    val id: String,
    val name: String,
)

@Serdeable
data class CreateCommentViewModel(
    val comment: String,
)
