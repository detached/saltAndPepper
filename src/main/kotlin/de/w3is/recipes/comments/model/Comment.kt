package de.w3is.recipes.comments.model

import de.w3is.recipes.recipes.model.RecipeId
import de.w3is.recipes.users.model.UserId
import java.time.Clock
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*

@JvmInline
value class CommentId(
    val id: UUID,
) {
    companion object {
        fun new() = CommentId(UUID.randomUUID())
    }
}

data class Comment(
    val id: CommentId,
    val userId: UserId,
    val recipeId: RecipeId,
    val text: String,
    val createdAt: Instant,
) {
    companion object {

        context(Clock)
        fun createNew(userId: UserId, recipeId: RecipeId, text: String) = Comment(
            id = CommentId.new(),
            userId = userId,
            recipeId = recipeId,
            text = text,
            createdAt = Instant.now(this@Clock).truncatedTo(ChronoUnit.MILLIS),
        )
    }
}
