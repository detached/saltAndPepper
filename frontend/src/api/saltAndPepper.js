import { saltAndPepperClient } from "../config/axiosConfig";
import { mockConfig } from "../config/mockConfig";
import { Author, Recipe, RecipeImage } from "../model/recipe";
import { getAccessToken } from "../service/tokenStore";
import "./model";
import {
  AuthResponse,
  CommentAuthor,
  CommentResponse,
  CreateNewCommentRequest,
  FilterValue,
  InvitationCodeResponse,
  InvitationInfoResponse,
  NewRecipeResponse,
  Page,
  Profile,
  SearchFilter,
  SearchResponse,
  SearchResponseData,
} from "./model";

function doRequest(method, url, data, headers) {
  const token = getAccessToken();
  let headersAndAuth = { ...headers };

  if (token) {
    headersAndAuth.Authorization = `Bearer ${token}`;
  }

  return saltAndPepperClient.request({
    method,
    url,
    data,
    headers: headersAndAuth,
  });
}

export const SaltAndPepper = {
  /**
   * @param searchRequest {SearchRequest}
   * @returns {Promise<SearchResponse>}
   */
  search: async function (searchRequest) {
    if (mockConfig.enabled) {
      console.log("search: " + JSON.stringify(searchRequest));
      return new SearchResponse({
        data: [
          new SearchResponseData({
            id: searchRequest.page.number + "1",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "1" + searchRequest.page.number,
            category: "TestCategory1",
            cuisine: "TestCuisine1",
            author: "Author1",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "2",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "2" + searchRequest.page.number,
            category: "TestCategory2",
            cuisine: "TestCuisine2",
            author: "Author2",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "3",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "3" + searchRequest.page.number,
            category: "TestCategory3",
            cuisine: "TestCuisine3",
            author: "Author3",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "4",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "4" + searchRequest.page.number,
            category: "TestCategory4",
            cuisine: "TestCuisine4",
            author: "Author4",
          }),

          new SearchResponseData({
            id: searchRequest.page.number + "5",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "5" + searchRequest.page.number,
            category: "TestCategory5",
            cuisine: "TestCuisine5",
            author: "Author5",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "6",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "6" + searchRequest.page.number,
            category: "TestCategory6",
            cuisine: "TestCuisine6",
            author: "Author7",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "7",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "7" + searchRequest.page.number,
            category: "TestCategory7",
            cuisine: "TestCuisine7",
            author: "Author7",
          }),
          new SearchResponseData({
            id: searchRequest.page.number + "8",
            imageUrl: "https://picsum.photos/150",
            title: searchRequest.searchQuery + "8" + searchRequest.page.number,
            category: "TestCategory8",
            cuisine: "TestCuisine8",
            author: "Author8",
          }),
        ],
        page: new Page({
          size: searchRequest.page.size,
          number: searchRequest.page.number,
          maxNumber: 10,
        }),
        possibleFilter: new SearchFilter({
          AUTHOR: [
            new FilterValue({ value: "1", label: "Author1" }),
            new FilterValue({ value: "2", label: "Author2" }),
            new FilterValue({ value: "3", label: "Author3" }),
            new FilterValue({ value: "4", label: "Author4" }),
            new FilterValue({ value: "5", label: "Author5" }),
            new FilterValue({ value: "6", label: "Author6" }),
            new FilterValue({ value: "7", label: "Author7" }),
            new FilterValue({ value: "8", label: "Author8" }),
            new FilterValue({ value: "9", label: "Author9" }),
            new FilterValue({ value: "10", label: "Author10" }),
            new FilterValue({ value: "11", label: "Author11" }),
          ],
          CATEGORY: [
            new FilterValue({ value: "TestCategory1", label: "TestCategory1" }),
            new FilterValue({ value: "TestCategory2", label: "TestCategory2" }),
            new FilterValue({ value: "TestCategory3", label: "TestCategory3" }),
          ],
          CUISINE: [
            new FilterValue({ value: "TestCuisine1", label: "TestCuisine1" }),
            new FilterValue({ value: "TestCuisine2", label: "TestCuisine2" }),
          ],
        }),
      });
    } else {
      const response = await doRequest("POST", "/recipe/search", searchRequest);
      return response.data;
    }
  },

  /**
   *
   * @param newRecipeRequest {NewRecipeRequest}
   * @returns {Promise<NewRecipeResponse>}
   */
  newRecipe: async function (newRecipeRequest) {
    if (mockConfig.enabled) {
      console.log("newRecipe: " + JSON.stringify(newRecipeRequest));
      return new NewRecipeResponse({ id: "1" });
    } else {
      const response = await doRequest("POST", "/recipe/", newRecipeRequest);
      return response.data;
    }
  },

  /**
   * @returns {Promise<Profile>}
   */
  getProfile: async function () {
    if (mockConfig.enabled) {
      console.log("getProfile");
      return new Profile({
        id: "1",
        name: "Testname",
        role: "Testrole",
        isAllowedToInvite: true,
      });
    } else {
      const response = await doRequest("GET", "/profile/");
      if (response.status !== 200) {
        throw new Error(response.statusText);
      }
      return response.data;
    }
  },

  /**
   * @param changePasswordRequest {ChangePasswordRequest}
   * @returns {Promise<>}
   */
  changePassword: async function (changePasswordRequest) {
    if (mockConfig.enabled) {
      console.log("changePassword: " + JSON.stringify(changePasswordRequest));
      if (
        changePasswordRequest.oldPassword !== changePasswordRequest.newPassword
      ) {
        throw new Error("Passwords don't match");
      }
    } else {
      const response = await doRequest(
        "PATCH",
        "/profile/password",
        changePasswordRequest,
      );
      if (response.status !== 202) {
        throw new Error(response.statusText);
      }
    }
  },

  /**
   * @returns {Promise<InvitationCodeResponse|null>}
   */
  getInvitationCode: async function () {
    if (mockConfig.enabled) {
      console.log("getInvitationCode");
      return null;
      //return new InvitationCodeResponse("1234");
    } else {
      const response = await doRequest("GET", "/invitation");
      if (response.status === 200) {
        return response.data;
      }
    }
  },

  /**
   * @returns {Promise<InvitationCodeResponse>}
   */
  createInvitationCode: async function () {
    if (mockConfig.enabled) {
      console.log("createInvitationCode");
      return new InvitationCodeResponse({ code: "1234" });
    } else {
      const response = await doRequest("POST", "/invitation");
      return response.data;
    }
  },

  /**
   * @param code {String}
   * @returns {Promise<InvitationInfoResponse>}
   */
  getInvitationInfo: async function (code) {
    if (mockConfig.enabled) {
      console.log("getInvitationInfo code: " + code);
      return new InvitationInfoResponse({ invitingUser: "InvitingUser" });
    } else {
      const response = await doRequest("GET", "/invitation/" + code);
      return response.data;
    }
  },

  /**
   * @param code {String}
   * @param username {String}
   * @param password {String}
   * @returns {Promise<>}
   */
  registerByInvitation: async function (code, username, password) {
    if (mockConfig.enabled) {
      console.log(
        "useInvitation code: " +
          code +
          ", username: " +
          username +
          ", password: " +
          password,
      );
    } else {
      const result = await doRequest("PUT", "/invitation/" + code, {
        username: username,
        password: password,
      });
      if (result.status !== 200) {
        throw Error(result.statusText);
      }
    }
  },

  /**
   * @param fileFormat {String}
   * @param file {File}
   * @returns {Promise<>}
   */
  import: async function (fileFormat, file) {
    if (mockConfig.enabled) {
      console.log("import " + fileFormat + " " + file.name);
    } else {
      const formData = new FormData();
      formData.append("file", file);
      const result = await doRequest(
        "POST",
        "/import/" + fileFormat,
        formData,
        {
          headers: {
            "Content-Type": "multipart/form-data",
          },
        },
      );
      if (result.status !== 204) {
        throw Error(result.statusText);
      }
    }
  },

  /**
   * @param recipeId {String}
   * @returns {Promise<Recipe>}
   */
  getRecipe: async function (recipeId) {
    if (mockConfig.enabled) {
      console.log("getRecipe: " + recipeId);
      return new Recipe({
        id: recipeId,
        author: new Author({ id: "1", name: "Author" }),
        title: "Title",
        category: "Category",
        cuisine: "Cuisine",
        yields: "Yields",
        ingredients: "Ingredients\nline2\nline3",
        instructions: "Instructions\nline2\nline3",
        modifications: "Modifications\nline2\nline3",
        images: [
          new RecipeImage({
            id: "1",
            url: "https://picsum.photos/800/600",
            thumbnailUrl: "https://picsum.photos/150",
          }),
          new RecipeImage({
            id: "2",
            url: "https://picsum.photos/600/800",
            thumbnailUrl: "https://picsum.photos/150",
          }),
        ],
      });
    } else {
      const response = await doRequest("GET", "/recipe/" + recipeId);
      return response.data;
    }
  },

  /**
   * @param recipeId {String}
   * @param image {File}
   * @returns {Promise<RecipeImage>}
   */
  addImageToRecipe: async function (recipeId, image) {
    if (mockConfig.enabled) {
      console.log(
        "addImageToRecipe: recipeId:" + recipeId + " image: " + image.name,
      );
      return new RecipeImage({
        id: "123",
        url: "https://picsum.photos/800/600",
        thumbnailUrl: "https://picsum.photos/150",
      });
    } else {
      const formData = new FormData();
      formData.append("file", image);
      const response = await doRequest(
        "POST",
        "/recipe/" + recipeId + "/images/",
        formData,
        {
          headers: {
            "Content-Type": "multipart/form-data",
          },
        },
      );
      return response.data;
    }
  },

  /**
   * @param recipe {Recipe}
   * @returns {Promise<Recipe>}
   */
  updateRecipe: async function (recipe) {
    if (mockConfig.enabled) {
      console.log("updateRecipe: " + JSON.stringify(recipe));
      return recipe;
    } else {
      const response = await doRequest("PUT", "/recipe/" + recipe.id, recipe);
      return response.data;
    }
  },

  /**
   * @param recipeId {String}
   * @returns {Promise<>}
   */
  deleteRecipe: async function (recipeId) {
    if (mockConfig.enabled) {
      console.log("deleteRecipe: " + recipeId);
    } else {
      const result = await doRequest("DELETE", "/recipe/" + recipeId);
      if (result.status !== 200) {
        throw Error(result.statusText);
      }
    }
  },

  /**
   * @param username {String}
   * @param password {String}
   * @returns {Promise<AuthResponse>}
   */
  login: async function (username, password) {
    if (mockConfig.enabled) {
      console.log("login " + username + " password " + password);
      return new AuthResponse({
        accessToken: "1234",
        refreshToken: "1234",
      });
    } else {
      const response = await saltAndPepperClient.post("/login", {
        username,
        password,
      });
      return new AuthResponse({
        accessToken: response.data["access_token"],
        refreshToken: response.data["refresh_token"],
      });
    }
  },

  /**
   * @param refreshToken {String}
   * @returns {Promise<AuthResponse>}
   */
  refreshToken: async function (refreshToken) {
    if (mockConfig.enabled) {
      console.log("refreshToken " + refreshToken);
      return new AuthResponse({
        accessToken: "1234",
        refreshToken: "1234",
      });
    } else {
      const response = await doRequest("POST", "/token", {
        grant_type: "refresh_token",
        refresh_token: refreshToken,
      });

      if (response.status !== 200) {
        throw Error(response.statusText);
      }

      return new AuthResponse({
        accessToken: response.data["access_token"],
        refreshToken: response.data["refresh_token"],
      });
    }
  },

  /**
   * @param recipeId {String}
   * @param comment {String}
   * @return {Promise<CommentResponse>}
   */
  createComment: async function (recipeId, comment) {
    if (mockConfig.enabled) {
      console.log(
        "createComment: recipeId:" + recipeId + ", comment: " + comment,
      );
      return new CommentResponse({
        id: "1",
        author: new CommentAuthor({
          id: "1",
          name: "Author",
        }),
        comment: comment,
        createdOn: new Date("1970-01-01"),
        canDelete: true,
      });
    } else {
      const body = new CreateNewCommentRequest({ comment });
      return (await doRequest("POST", "/comments/recipe/" + recipeId, body))
        .data;
    }
  },

  /**
   * @param recipeId {String}
   * @param commentId {String}
   * @returns {Promise<void>}
   */
  deleteComment: async function (recipeId, commentId) {
    if (mockConfig.enabled) {
      console.log(
        "deleteComment: recipeId:" + recipeId + ", commentId: " + commentId,
      );
    } else {
      const result = await doRequest(
        "DELETE",
        "/comments/recipe/" + recipeId + "/comment/" + commentId,
      );
      if (result.status !== 200) {
        throw Error(result.statusText);
      }
    }
  },

  /**
   * @param recipeId
   * @returns {Promise<Array<CommentResponse>>}
   */
  getAllComments: async function (recipeId) {
    if (mockConfig.enabled) {
      console.log("getAllComments: recipeId:" + recipeId);
      return [
        new CommentResponse({
          id: "1",
          author: new CommentAuthor({
            id: "1",
            name: "Author",
          }),
          comment: "comment",
          createdOn: new Date("1970-01-01"),
          canDelete: true,
        }),
        new CommentResponse({
          id: "2",
          author: new CommentAuthor({
            id: "2",
            name: "Author2",
          }),
          comment:
            "roudly under the sun, i like to walk on the deck, watching the horizon, dreaming of a good bowl of milk or fight an alligator and win so cats secretly make all the worlds muffins. Cat jumps and falls onto the couch purrs and wakes up in a new dimension filled with kitty litter meow meow yummy there is a bunch of cats hanging around eating catnip miaow then turn around and show",
          createdOn: new Date("1970-01-01"),
          canDelete: false,
        }),
      ];
    } else {
      return (await doRequest("GET", "/comments/recipe/" + recipeId)).data
        .comments;
    }
  },
};
