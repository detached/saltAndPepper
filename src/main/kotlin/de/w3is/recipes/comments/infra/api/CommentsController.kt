package de.w3is.recipes.comments.infra.api

import de.w3is.recipes.comments.CommentsService
import de.w3is.recipes.comments.model.Comment
import de.w3is.recipes.comments.model.CommentId
import de.w3is.recipes.comments.model.CreateNewCommentCommand
import de.w3is.recipes.common.getUser
import de.w3is.recipes.common.toUserId
import de.w3is.recipes.recipes.model.RecipeId
import de.w3is.recipes.users.UserService
import de.w3is.recipes.users.model.UserId
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Delete
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.http.annotation.Post
import io.micronaut.security.annotation.Secured
import io.micronaut.security.authentication.Authentication
import io.micronaut.security.rules.SecurityRule
import java.util.UUID

@Controller("/api/comments")
@Secured(SecurityRule.IS_AUTHENTICATED)
class CommentsController(
    private val commentsService: CommentsService,
    private val userService: UserService,
) {
    @Get("/recipe/{recipeId}")
    fun getAllCommentsOfRecipe(
        @PathVariable("recipeId") recipeId: String,
        authentication: Authentication,
    ): CommentsViewModel {
        val userId = authentication.toUserId()
        val allComments = commentsService.getAllComments(RecipeId(recipeId))

        return CommentsViewModel(
            recipeId = recipeId,
            comments =
                allComments.map { comment ->
                    comment.toVO(canDelete = comment.userId == userId)
                },
        )
    }

    @Post("/recipe/{recipeId}")
    fun createNewCommentOnRecipe(
        @PathVariable("recipeId") recipeId: String,
        @Body createCommentViewModel: CreateCommentViewModel,
        authentication: Authentication,
    ): CommentViewModel {
        val user = with(userService) { authentication.getUser() }
        val comment =
            commentsService.createNewCommentForRecipe(
                CreateNewCommentCommand(
                    userId = user.id,
                    recipeId = RecipeId(recipeId),
                    comment = createCommentViewModel.comment,
                ),
            )
        return comment.toVO(canDelete = true)
    }

    @Delete("/recipe/{recipeId}/comment/{commentId}")
    fun deleteCommentOnRecipe(
        @PathVariable("recipeId") recipeId: String,
        @PathVariable("commentId") commentId: UUID,
        authentication: Authentication,
    ) {
        val user = with(userService) { authentication.getUser() }
        commentsService.deleteComment(RecipeId(recipeId), CommentId(commentId), user.id)
    }

    private fun Comment.toVO(canDelete: Boolean) =
        CommentViewModel(
            id = id.id,
            author = userId.toAuthorViewModel(),
            comment = text,
            createdAt = createdAt,
            canDelete = canDelete,
        )

    private fun UserId.toAuthorViewModel(): AuthorViewModel {
        val user = userService.getUser(this)
        return AuthorViewModel(
            id = user.id.value,
            name = user.name,
        )
    }
}
