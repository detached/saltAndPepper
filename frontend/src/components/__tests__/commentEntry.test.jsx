import { render } from "@testing-library/react";
import CommentEntry from "../commentEntry";
import { CommentAuthor, CommentResponse } from "../../api/model";

describe("commentEntry snapshot tests", () => {
  test("with delete button", () => {
    const comment = new CommentResponse({
      id: "commentId1",
      author: new CommentAuthor({
        id: "authorId1",
        name: "Author Name",
      }),
      comment: "That is a comment",
      createdOn: new Date("1970-01-01"),
      canDelete: true,
    });

    const { container } = render(<CommentEntry comment={comment} />);

    expect(container).toMatchSnapshot();
  });

  test("without delete button", () => {
    const comment = new CommentResponse({
      id: "commentId1",
      author: new CommentAuthor({
        id: "authorId1",
        name: "Author Name",
      }),
      comment: "That is a comment",
      createdOn: new Date("1970-01-01"),
      canDelete: false,
    });

    const { container } = render(<CommentEntry comment={comment} />);

    expect(container).toMatchSnapshot();
  });
});
