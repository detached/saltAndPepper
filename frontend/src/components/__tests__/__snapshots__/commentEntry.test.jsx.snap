// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`commentEntry snapshot tests > with delete button 1`] = `
<div>
  <li
    class="pure-g comment-entry-comment"
  >
    <div
      class="comment-entry-left"
    >
      <svg
        class="comment-entry-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 7.1 5.8 12 12 12 2.4 0 4.9-.7 7.1-2.4L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64zm16 352c0 8.8-7.2 16-16 16H288l-12.8 9.6L208 428v-60H64c-8.8 0-16-7.2-16-16V64c0-8.8 7.2-16 16-16h384c8.8 0 16 7.2 16 16v288z"
        />
      </svg>
    </div>
    <div
      class=""
    >
      <div
        class="comment-entry-content-title"
      >
        <div
          class="comment-entry-author"
        >
          Author Name
        </div>
        <div
          class="comment-entry-date"
        >
          1.1.1970
        </div>
        <svg
          class="comment-entry-control"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M268 416h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12zM432 80h-82.41l-34-56.7A48 48 0 0 0 274.41 0H173.59a48 48 0 0 0-41.16 23.3L98.41 80H16A16 16 0 0 0 0 96v16a16 16 0 0 0 16 16h16v336a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128h16a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM171.84 50.91A6 6 0 0 1 177 48h94a6 6 0 0 1 5.15 2.91L293.61 80H154.39zM368 464H80V128h288zm-212-48h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12z"
          />
        </svg>
      </div>
      <p
        class="comment-entry-text"
      >
        That is a comment
      </p>
    </div>
  </li>
</div>
`;

exports[`commentEntry snapshot tests > without delete button 1`] = `
<div>
  <li
    class="pure-g comment-entry-comment"
  >
    <div
      class="comment-entry-left"
    >
      <svg
        class="comment-entry-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 7.1 5.8 12 12 12 2.4 0 4.9-.7 7.1-2.4L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64zm16 352c0 8.8-7.2 16-16 16H288l-12.8 9.6L208 428v-60H64c-8.8 0-16-7.2-16-16V64c0-8.8 7.2-16 16-16h384c8.8 0 16 7.2 16 16v288z"
        />
      </svg>
    </div>
    <div
      class=""
    >
      <div
        class="comment-entry-content-title"
      >
        <div
          class="comment-entry-author"
        >
          Author Name
        </div>
        <div
          class="comment-entry-date"
        >
          1.1.1970
        </div>
      </div>
      <p
        class="comment-entry-text"
      >
        That is a comment
      </p>
    </div>
  </li>
</div>
`;
